### PR TITLE
feat: add MariaDB Vector backend

### PR DESCRIPTION
**Pull request by seulement55**
# Pull Request Checklist

- [x] **Target branch:** Verify that the pull request targets the `dev` branch. **PRs targeting `main` will be immediately closed.**
- [x] **Description:** Provide a concise description of the changes made in this pull request down below.
- [x] **Changelog:** Ensure a changelog entry following the format of [Keep a Changelog](https://keepachangelog.com/) is added at the bottom of the PR description.
- [x] **Documentation:** Add docs in [Open WebUI Docs Repository](https://github.com/open-webui/docs). Document user-facing behavior, environment variables, public APIs/interfaces, or deployment steps.
- [x] **Dependencies:** Are there any new or upgraded dependencies? If so, explain why, update the changelog/docs, and include any compatibility notes. Actually run the code/function that uses updated library to ensure it doesn't crash.
- [x] **Testing:** Perform manual tests to **verify the implemented fix/feature works as intended AND does not break any other functionality**. Include reproducible steps to demonstrate the issue before the fix. Test edge cases (URL encoding, HTML entities, types). Take this as an opportunity to **make screenshots of the feature/fix and include them in the PR description**.
- [x] **Agentic AI Code:** Confirm this Pull Request is **not written by any AI Agent** or has at least **gone through additional human review AND manual testing**. If any AI Agent is the co-author of this PR, it may lead to immediate closure of the PR.
- [x] **Code review:** Have you performed a self-review of your code, addressing any coding standard issues and ensuring adherence to the project's coding standards?
- [x] **Design & Architecture:** Prefer smart defaults over adding new settings; use local state for ephemeral UI logic. Open a Discussion for major architectural or UX changes.
- [x] **Git Hygiene:** Keep PRs atomic (one logical change). Clean up commits and rebase on `dev` to ensure no unrelated commits (e.g. from `main`) are included. Push updates to the existing PR branch instead of closing and reopening.
- [x] **Title Prefix:** To clearly categorize this pull request, prefix the pull request title using one of the following:
  - **feat**: Introduces a new feature or enhancement to the codebase

---

# Description

This PR adds **MariaDB Vector** as a supported vector database backend in Open-WebUI, enabling `VECTOR_DB=mariadb-vector` deployments. The implementation uses the **official MariaDB connector** (`mariadb+mariadbconnector://...`) to ensure correct parameter binding of `VECTOR(n)` values (float32 binary payload), and provides configurable index and pooling parameters.

This work addresses the request to support MariaDB as a vector database backend (see discussion [https://github.com/open-webui/open-webui/discussions/21363](https://github.com/open-webui/open-webui/discussions/21363)).

## Key changes

Adds a new `mariadb-vector` vector database backend, including configuration/env support and factory wiring, plus a MariaDB Vector client implementation that initializes the required schema and performs insert/upsert/search operations using the official MariaDB connector for correct `VECTOR(n)` float32 binding. Updates container and Python dependencies to include MariaDB driver support, and exposes tuning knobs for vector length, distance strategy, HNSW `M`, and connection pooling.

## Schema and compatibility notes

* MariaDB Vector imposes constraints on primary keys used with `VECTOR INDEX`. This backend uses `id VARCHAR(254) CHARACTER SET ascii COLLATE ascii_bin PRIMARY KEY` to ensure the PK stays under the 256-byte limit and comparisons are stable.
* The backend checks the existing DDL for `VECTOR(n)` mismatch and fails fast if the configured dimension differs from the stored column dimension.

## Dependencies

New dependency added:

* `mariadb==1.1.14` (Python driver), plus `libmariadb-dev` in the container image.

Rationale:

* MariaDB Vector requires correct binding of `VECTOR(n)` values as float32 binary payloads; the official MariaDB connector provides the expected driver behavior and qmark paramstyle.

Compatibility notes:

* `MARIADB_VECTOR_DB_URL` **must** use `mariadb+mariadbconnector://...` (official driver).
* MariaDB server must support `VECTOR` and `VECTOR INDEX` features.

### Manual test (repro steps)

**1) Build image**

```bash
docker build -f Dockerfile -t open-webui:0.9.0-dev .
```

**2) Start Open-WebUI (Postgres main DB + MariaDB Vector DB)**

```bash
docker run -d \
  --network host \
  -e OLLAMA_BASE_URL="http://ollama-host.com:11434" \
  -e AIOHTTP_CLIENT_TIMEOUT=3600 \
  \
  -e ENABLE_PERSISTENT_CONFIG=False \
  -e DEFAULT_MODEL_PARAMS='{"num_ctx":32768,"temperature":0.1,"top_p":0.9}' \
  \
  -e RAG_EMBEDDING_ENGINE=ollama \
  -e RAG_OLLAMA_BASE_URL="http://ollama-host.com:11434" \
  -e RAG_EMBEDDING_MODEL=nomic-embed-text \
  -e RAG_EMBEDDING_BATCH_SIZE=64 \
  -e RAG_TEXT_SPLITTER=token \
  -e RAG_SYSTEM_CONTEXT=True \
  -e RAG_TOP_K=10 \
  -e RAG_TOP_K_RERANKER=10 \
  \
  -e CHUNK_SIZE=1024 \
  -e CHUNK_OVERLAP=128 \
  -e CHUNK_MIN_SIZE_TARGET=640 \
  -e ENABLE_ASYNC_EMBEDDING=true \
  -e RAG_EMBEDDING_CONCURRENT_REQUESTS=8 \
  -e ENABLE_RAG_HYBRID_SEARCH=true \
  \
  -e DATABASE_URL="postgres://app:app@127.0.0.1:5432/openwebui" \
  -e VECTOR_DB="mariadb-vector" \
  -e MARIADB_VECTOR_DB_URL="mariadb+mariadbconnector://app:app@127.0.0.1:3306/openwebui" \
  -e MARIADB_VECTOR_INITIALIZE_MAX_VECTOR_LENGTH=768 \
  -e MARIADB_VECTOR_DISTANCE_STRATEGY=cosine \
  -e MARIADB_VECTOR_INDEX_M=12 \
  \
  -v open-webui:/app/backend/data \
  --name open-webui \
  --restart always \
  open-webui:0.9.0-dev
```

**3) Configure RAG in the UI**
Follow the official RAG tutorial: [https://docs.openwebui.com/tutorials/tips/rag-tutorial/#setup](https://docs.openwebui.com/tutorials/tips/rag-tutorial/#setup)

**4) Validate ingestion + retrieval**

* Upload Open-WebUI docs as a Knowledge Base
* Run RAG queries and verify that sources are returned and answers reference the correct KB chunks

### Observed results

When comparing similarity search results between `mariadb-vector` and `pgvector` using Open-WebUI documentation as the knowledge base and `llama3.1:8b` as the base model, retrieval outcomes were similar in most cases, with no major differences observed.

## Documentation

* [x] I will add documentation to [https://github.com/open-webui/docs](https://github.com/open-webui/docs) covering: https://github.com/open-webui/docs/pull/1125

  * `VECTOR_DB=mariadb-vector` setup
  * `MARIADB_VECTOR_*` environment variables
  * required driver scheme: `mariadb+mariadbconnector://...`

## Screenshots

* [x] Knowledge Base ingestion
image

* [x] Example RAG answers with sources.
image

---

# Changelog Entry

### Description

* Add MariaDB Vector as a supported vector database backend for RAG similarity search, allowing Open-WebUI deployments to store and query embeddings using MariaDB’s `VECTOR` and `VECTOR INDEX` features.

### Added

* Added `mariadb-vector` as a vector database backend for RAG retrieval, including new `MARIADB_VECTOR_*` configuration options and a MariaDB Vector client implementation for insert/upsert/search. Requires the official MariaDB connector scheme (`mariadb+mariadbconnector://`) for correct `VECTOR(n)` binding.

---

### Contributor License Agreement

By submitting this pull request, I confirm that I have read and fully agree to the [Contributor License Agreement (CLA)](https://github.com/open-webui/open-webui/blob/main/CONTRIBUTOR_LICENSE_AGREEMENT), and I am providing my contributions under its terms.

> [!NOTE]
> Deleting the CLA section will lead to immediate closure of your PR and it will not be merged in.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -127,6 +127,7 @@ RUN chown -R $UID:$GID /app $HOME
 RUN apt-get update && \
     apt-get install -y --no-install-recommends \
     git build-essential pandoc gcc netcat-openbsd curl jq \
+    libmariadb-dev \
```
Comment (tjbck) [inline Dockerfile:130]: How big is this package?

**Comment by seulement55** [comment]
Hi @tjbck 
I noticed this PR was marked as draft. Could you please let me know what needs to be improved or updated?
I’d be happy to address any feedback. 

Thank you!

**Comment by seulement55** [comment]
hi @tjbck 

I have created a PR to update the documentation for this feature https://github.com/open-webui/docs/pull/1125  and have already tested the changes in our lab environment. So, I'm marking this PR as `ready for review` and would greatly appreciate your feedback. Thank you!

File: Dockerfile (hunk 1 above)
Comment (seulement55) [reply Dockerfile:130]: `libmariadb-dev` itself is relatively small (roughly **~0.27 MB** to download and **~1 MB** installed on Ubuntu amd64). Reference: [https://packages.ubuntu.com/noble-updates/libmariadb-dev](https://packages.ubuntu.com/noble-updates/libmariadb-dev)

Below are the Docker image sizes before and after installing this package:

```bash
user@ai:[debug_open-webui] [] $ docker images

IMAGE                                 ID             DISK USAGE   CONTENT SIZE   EXTRA
ghcr.io/open-webui/open-webui:0.8.5   57ce4ee1dad9       4.64GB             0B
open-webui:0.9.0-dev                  4d6934e6042e       4.66GB             0B
```

This is an increase of approximately **~0.02 GB (~20 MB)**. This delta should be interpreted as the combined layer impact of adding `libmariadb-dev` and its dependencies.

**Comment by tjbck** [comment]
Thanks!
